### PR TITLE
feat(resource)!: require DynamicClient client

### PR DESCRIPTION
**Pull request by rnetser**
## Summary
- Make `client: DynamicClient` mandatory on `Resource` / `NamespacedResource` / `Event` APIs (`__init__`, `.get()`, `get_all_cluster_resources`, `Event.get` / `delete_events` / `list`).
- Remove deprecated `dyn_client` and `config_file` / `config_dict` / `context` from Resource APIs (still available on `get_client()`).
- Reject legacy kwargs and extra positionals after `client`; raise clear `TypeError`s.
- Update `AGENTS.md`, validation examples, and tests.

**BREAKING CHANGE** — callers must pass an explicit client from `get_client()` / `get_client(fake=True)`. Version bump to next major is intended to be done manually at release.

## Test plan
- [ ] `uv run --group tests pytest tests/test_resource.py::TestResource::test_client_is_required tests/test_resource.py::TestResource::test_removed_client_kwargs_rejected tests/test_resource.py::TestResource::test_event_client_is_required`
- [ ] `prek run && uv run --group tests pytest`
- [ ] Confirm CI green
- [ ] Regenerate `docs/` with docsfy after merge/release (hand-edits forbidden)

Made with [Cursor](https://cursor.com)

## Summary by CodeRabbit

* **Breaking Changes**
  * Resource creation, retrieval, and cluster-wide listing now require an explicit client.
  * Event retrieval, deletion, and listing require an explicit client.
  * Legacy client configuration options and unsupported positional arguments are no longer accepted.

* **Documentation**
  * Updated examples and guidance to demonstrate explicit client usage and reusable fake clients.

* **Tests**
  * Expanded coverage for required clients, rejected legacy arguments, and explicit-client workflows.

### CODE REVIEW

File: ocp_resources/resource.py [hunk 1]
```diff
@@ -615,17 +627,14 @@ class ApiVersion:
 
     def __init__(
         self,
-        client: DynamicClient | None = None,  # TODO: make mandatory in the next major release
+        client: DynamicClient,
```
[the remark this answers is not in the document]
Comment (rnetser) [reply ocp_resources/resource.py:637]: Addressed: added `-> None` return annotation on `Resource.__init__` and normalized Google-style Args/Raises docstring (no typed param names; summary ends with period).

---
*Assisted-by: PI (cursor:cursor-grok-4.5-high)*

File: ocp_resources/resource.py [hunk 2]
```diff
@@ -1625,12 +1604,12 @@ class NamespacedResource(Resource):
 
     def __init__(
         self,
+        client: DynamicClient,
```
[the remark this answers is not in the document]
Comment (rnetser) [reply ocp_resources/resource.py:1620]: Addressed: added `-> None` return annotation on `NamespacedResource.__init__`.

---
*Assisted-by: PI (cursor:cursor-grok-4.5-high)*

File: tests/test_resource.py [hunk 3]
```diff
@@ -96,6 +97,72 @@ def test_get_all_cluster_resources(self, fake_client):
             if _resources:
                 break
 
+    def test_client_is_required(self, fake_client):
```
[the remark this answers is not in the document]
Comment (rnetser) [reply tests/test_resource.py]: Addressed: annotated `fake_client: FakeDynamicClient` and `-> None` on `test_client_is_required`.

---
*Assisted-by: PI (cursor:cursor-grok-4.5-high)*

File: tests/test_resource.py [hunk 4]
```diff
@@ -96,6 +97,72 @@ def test_get_all_cluster_resources(self, fake_client):
...
+            list(Resource.get_all_cluster_resources())
+
+        with pytest.raises(TypeError, match="client"):
+            list(Resource.get_all_cluster_resources(client=None))  # type: ignore[arg-type]
+
+        # Valid construction still works with an explicit client
+        pod = Pod(client=fake_client, name=BASE_POD_NAME, namespace="default", containers=POD_CONTAINERS)
+        assert pod.client is fake_client
+
+    def test_removed_client_kwargs_rejected(self, fake_client):
```
[the remark this answers is not in the document]
Comment (rnetser) [reply tests/test_resource.py]: Addressed: annotated `fake_client: FakeDynamicClient` and `-> None` on `test_removed_client_kwargs_rejected`.

---
*Assisted-by: PI (cursor:cursor-grok-4.5-high)*

File: tests/test_resource.py [hunk 5]
```diff
@@ -96,6 +97,72 @@ def test_get_all_cluster_resources(self, fake_client):
...
+        with pytest.raises(TypeError, match="config_file"):
+            Pod(
+                client=fake_client,
+                name=BASE_POD_NAME,
+                namespace="default",
+                containers=POD_CONTAINERS,
+                config_file="/tmp/kubeconfig",
+            )
+
+    def test_event_client_is_required(self):
```
[the remark this answers is not in the document]
Comment (rnetser) [reply tests/test_resource.py]: Addressed: added `-> None` return annotation on `test_event_client_is_required`.

---
*Assisted-by: PI (cursor:cursor-grok-4.5-high)*

**Comment by rnetser** [comment]
/qodo

The following review comments were reviewed and a decision was made:

### `ocp_resources/resource.py:630` (qodo rule violation) — Resource.__init__ lacks return type
> Addressed: Addressed: added `-> None` return annotation on `Resource.__init__` and normalized Google-style Args/Raises docstring (no typed param names; summary ends with period).
### `ocp_resources/resource.py:1607` (qodo rule violation) — NamespacedResource.__init__ lacks return type
> Addressed: Addressed: added `-> None` return annotation on `NamespacedResource.__init__`.
### `tests/test_resource.py:100` (qodo rule violation) — test_client_is_required is untyped
> Addressed: Addressed: annotated `fake_client: FakeDynamicClient` and `-> None` on `test_client_is_required`.
### `tests/test_resource.py:123` (qodo rule violation) — test_removed_client_kwargs_rejected is untyped
> Addressed: Addressed: annotated `fake_client: FakeDynamicClient` and `-> None` on `test_removed_client_kwargs_rejected`.
### `tests/test_resource.py:153` (qodo rule violation) — test_event_client_is_required lacks return type
> Addressed: Addressed: added `-> None` return annotation on `test_event_client_is_required`.
### `ocp_resources/resource.py:199` (qodo rule violation) — Helper docstring omits sections
> Addressed: Addressed: expanded `_reject_removed_client_kwargs` docstring with Google-style Args and Raises sections.
### `ocp_resources/resource.py:654` (qodo rule violation) — Constructor docstring uses typed names
> Addressed: Addressed: normalized `Resource.__init__` docstring to Google style (`name:` not `name (str):`), summary ends with period, and added Raises section.

---
*Assisted-by: PI (cursor:cursor-grok-4.5-high)*

File: ocp_resources/event.py [hunk 6]
```diff
@@ -16,29 +15,10 @@ class Event:
...
-        resolved = client or dyn_client
-        assert resolved is not None, "Either 'client' or 'dyn_client' must be provided"
-        return resolved
-
     @classmethod
     def get(
         cls,
-        client: DynamicClient | None = None,  # TODO: make mandatory in the next major release
-        dyn_client: DynamicClient | None = None,  # TODO: remove in the next major release
+        client: DynamicClient,
```
[the remark this answers is not in the document]
Comment (rnetser) [reply ocp_resources/event.py:21]: Thanks — but we're intentionally declining this. This PR is already a breaking change (making `client` mandatory). Converting `Event.get`/`Event.list`/`Event.delete_events` to keyword-only would layer an additional breaking change onto the same release, which we want to avoid. The positional-binding edge case can be revisited separately if needed.

File: tests/test_resource.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (rnetser) [reply tests/test_resource.py]: Already addressed. `test_client_is_required` and `test_removed_client_kwargs_rejected` are annotated `(self, fake_client: FakeDynamicClient) -> None`, and `test_event_client_is_required` has `-> None`.

File: tests/test_resource.py [hunk 7]
```diff
@@ -96,6 +97,72 @@ def test_get_all_cluster_resources(self, fake_client):
...
+        # Valid construction still works with an explicit client
+        pod = Pod(client=fake_client, name=BASE_POD_NAME, namespace="default", containers=POD_CONTAINERS)
+        assert pod.client is fake_client
+
+    def test_removed_client_kwargs_rejected(self, fake_client):
+        with pytest.raises(TypeError, match="Unsupported argument"):
+            list(Namespace.get(client=fake_client, dyn_client=fake_client))
+
+        with pytest.raises(TypeError, match="Unsupported argument"):
+            list(Namespace.get(client=fake_client, config_file="/tmp/kubeconfig"))
```
[the remark this answers is not in the document]
Comment (rnetser) [reply tests/test_resource.py]: Already addressed. Both `/tmp/kubeconfig` literals were replaced with the non-temporary placeholder `"kubeconfig"`, clearing Ruff S108.
